core/src/main/java/io/lionweb/lioncore/java/utils/LanguageValidator.java

```diff
@@ -15,6 +15,63 @@ public static void ensureIsValid(Language language) {
     }
   }
 
+  private void validateEnumeration(ValidationResult result, Enumeration enumeration) {
+    enumeration
+        .getLiterals()
+        .forEach(
+            (EnumerationLiteral lit) ->
+                result.checkForError(lit.getName() == null, "Simple name not set", lit));
+    validateNamesAreUnique(enumeration.getLiterals(), result);
+  }
+
+  private void validateClassifier(ValidationResult result, Classifier<?> classifier) {
+    classifier
+        .getFeatures()
+        .forEach(
+            (Feature feature) ->
+                result
+                    .checkForError(feature.getName() == null, "Simple name not set", feature)
+                    .checkForError(feature.getContainer() == null, "Container not set", feature)
+                    .checkForError(
+                        feature.getContainer() != null && feature.getContainer() != classifier,
+                        "Features container not set correctly",
+                        feature));
+    validateNamesAreUnique(classifier.getFeatures(), result);
+  }
+
+  private void validateConcept(ValidationResult result, Concept concept) {
+    checkAncestors(concept, result);
+    result.checkForError(
+        concept.getImplemented().size() != concept.getImplemented().stream().distinct().count(),
+        "The same interface has been implemented multiple times",
+        concept);
+  }
+
+  private void validateStructuralDataType(
+      ValidationResult result, StructuredDataType structuredDataType) {
+    Set<StructuredDataType> visited = new HashSet<>();
+    Stack<StructuredDataType> toVisit = new Stack<>();
+    toVisit.add(structuredDataType);
+
+    while (!toVisit.isEmpty()) {
```

Comment:
This algorithm begs to be extracted out to a separate place, so that you can test its functionality thoroughly, and just have to call it here.

Reply:
Right. I have created a public method (`LanguageValidator.isCircular`) which is invoked from here and added a couple of tests for it